File: Sources/ReminderCore/Store.swift
import Foundation

/// The on-disk shape of everything the app persists.
public struct AppData: Codable, Equatable, Sendable {
    /// Bumped when the on-disk format changes, so migrations have something to
    /// branch on.
    public var schemaVersion: Int
    public var reminders: [Reminder]
    public var settings: Settings
    public var events: [ReminderEvent]

    public static let currentSchemaVersion = 1

    public init(
        schemaVersion: Int = AppData.currentSchemaVersion,
        reminders: [Reminder] = [],
        settings: Settings = Settings(),
        events: [ReminderEvent] = []
    ) {
        self.schemaVersion = schemaVersion
        self.reminders = reminders
        self.settings = settings
        self.events = events
    }
}

/// Abstracts the filesystem so tests can run against a temp directory.
public protocol DataStoring: AnyObject {
    func load() throws -> AppData
    func save(_ data: AppData) throws
}

/// Persists `AppData` as pretty-printed JSON in a single local file.
///
/// Everything stays on this machine — there is no network path out of this
/// type, by design.
public final class FileDataStore: DataStoring, @unchecked Sendable {
    public let fileURL: URL
    private let queue = DispatchQueue(label: "com.reminder.filedatastore")

    /// The default location: `~/Library/Application Support/Reminder/data.json`.
    public static func defaultFileURL(
        fileManager: FileManager = .default
    ) throws -> URL {
        let base = try fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let directory = base.appendingPathComponent("Reminder", isDirectory: true)
        try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory.appendingPathComponent("data.json")
    }

    public init(fileURL: URL) {
        self.fileURL = fileURL
    }

    public convenience init(fileManager: FileManager = .default) throws {
        self.init(fileURL: try FileDataStore.defaultFileURL(fileManager: fileManager))
    }

    private static func makeEncoder() -> JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        // Numeric timestamps rather than ISO8601: ISO8601 drops fractional
        // seconds, which would round `lastFiredAt` on every save/load cycle and
        // let interval schedules drift.
        encoder.dateEncodingStrategy = .secondsSince1970
        return encoder
    }

    private static func makeDecoder() -> JSONDecoder {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .secondsSince1970
        return decoder
    }

    public func load() throws -> AppData {
        try queue.sync {
            guard FileManager.default.fileExists(atPath: fileURL.path) else {
                return AppData(reminders: DefaultReminders.starterSet())
            }
            let raw = try Data(contentsOf: fileURL)
            guard !raw.isEmpty else {
                return AppData(reminders: DefaultReminders.starterSet())
            }
            return try Self.makeDecoder().decode(AppData.self, from: raw)
        }
    }

    public func save(_ data: AppData) throws {
        try queue.sync {
            let encoded = try Self.makeEncoder().encode(data)
            let directory = fileURL.deletingLastPathComponent()
            try FileManager.default.createDirectory(
                at: directory, withIntermediateDirectories: true
            )
            // Atomic write: a crash mid-save must not leave a truncated file
            // that loses every reminder the user configured.
            try encoded.write(to: fileURL, options: .atomic)
        }
    }
}

/// An in-memory store for tests and previews.
public final class InMemoryDataStore: DataStoring, @unchecked Sendable {
    public private(set) var data: AppData
    public private(set) var saveCount = 0

    public init(data: AppData = AppData()) {
        self.data = data
    }

    public func load() throws -> AppData { data }

    public func save(_ data: AppData) throws {
        self.data = data
        saveCount += 1
    }
}

/// The reminders a brand-new install starts with.
public enum DefaultReminders {
    /// Pressure relief and movement prompts, plus a couple of general examples
    /// so the concept of priority tiers is visible immediately.
    public static func starterSet(now: Date = Date()) -> [Reminder] {
        [
            Reminder(
                title: "Tilt Back",
                message: "Tilt your chair back for 5 minutes. Stop working and listen to calming music.",
                schedule: .interval(minutes: 60),
                priority: .critical,
                symbolName: "figure.seated.side",
                activityDurationSeconds: 5 * 60,
                createdAt: now
            ),
            Reminder(
                title: "Weight Shift",
                message: "Activate your glutes and redistribute your weight.",
                schedule: .interval(minutes: 20),
                priority: .subtle,
                symbolName: "arrow.up.and.down.circle",
                createdAt: now
            ),
            Reminder(
                title: "Drink Water",
                message: "Have a drink of water.",
                schedule: .interval(minutes: 60),
                priority: .normal,
                symbolName: "drop.fill",
                createdAt: now
            ),
            Reminder(
                title: "Stretch & Range of Motion",
                message: "Run through your physio stretches.",
                schedule: .dailyAt(hour: 17, minute: 0, dayInterval: 2),
                priority: .important,
                symbolName: "figure.flexibility",
                activityDurationSeconds: 10 * 60,
                isEnabledDefault: false,
                createdAt: now
            ),
        ]
    }
}

private extension Reminder {
    /// Convenience for the starter set: some examples ship switched off so the
    /// user opts in rather than being surprised by them.
    init(
        title: String,
        message: String,
        schedule: Schedule,
        priority: Priority,
        symbolName: String,
        activityDurationSeconds: Int? = nil,
        isEnabledDefault: Bool,
        createdAt: Date
    ) {
        self.init(
            title: title,
            message: message,
            schedule: schedule,
            priority: priority,
            isEnabled: isEnabledDefault,
            symbolName: symbolName,
            activityDurationSeconds: activityDurationSeconds,
            createdAt: createdAt
        )
    }
}

File: Windows/Pauselet.Core/Store.cs
using NodaTime;

namespace Pauselet.Core;

/// <summary>The on-disk shape of everything the app persists.</summary>
public sealed record AppData
{
    /// <summary>
    /// Bumped when the on-disk format changes, so migrations have something to
    /// branch on.
    /// </summary>
    public int SchemaVersion { get; init; } = CurrentSchemaVersion;
    public IReadOnlyList<Reminder> Reminders { get; init; } = [];
    public Settings Settings { get; init; } = new();
    public IReadOnlyList<ReminderEvent> Events { get; init; } = [];

    public const int CurrentSchemaVersion = 1;

    // List-valued properties need element-wise comparison; the synthesized
    // record equality would compare list references.
    public bool Equals(AppData? other) =>
        other is not null
        && SchemaVersion == other.SchemaVersion
        && Reminders.SequenceEqual(other.Reminders)
        && Settings == other.Settings
        && Events.SequenceEqual(other.Events);

    public override int GetHashCode() =>
        HashCode.Combine(SchemaVersion, Reminders.Count, Settings, Events.Count);
}

/// <summary>Abstracts the filesystem so tests can run against a temp directory.</summary>
public interface IDataStoring
{
    AppData Load();
    void Save(AppData data);
    /// <summary>
    /// Whether anything has actually been written yet. Lets the engine tell a
    /// genuine first launch from a returning one, so it can persist the starter
    /// set once rather than reseeding timing anchors on every launch.
    /// </summary>
    bool HasPersistedData { get; }
}

/// <summary>
/// Persists <c>AppData</c> as pretty-printed JSON in a single local file.
///
/// Everything stays on this machine — there is no network path out of this
/// type, by design. The file is byte-compatible with the Mac app's, so copying
/// it between machines is a supported migration path.
/// </summary>
public sealed class FileDataStore : IDataStoring
{
    public string FilePath { get; }
    private readonly object _gate = new();

    /// <summary>The folder the app stores data in, under the roaming app-data directory.</summary>
    internal const string DirectoryName = "Pauselet";

    /// <summary>
    /// The default location: <c>%APPDATA%\Pauselet\data.json</c> on Windows.
    /// (The Mac app's legacy-directory migration has no counterpart here —
    /// there has never been a Windows install under the old name.)
    /// </summary>
    public static string DefaultFilePath()
    {
        var baseDirectory = Environment.GetFolderPath(
            Environment.SpecialFolder.ApplicationData,
            Environment.SpecialFolderOption.Create
        );
        var directory = Path.Combine(baseDirectory, DirectoryName);
        Directory.CreateDirectory(directory);
        return Path.Combine(directory, "data.json");
    }

    public FileDataStore(string filePath)
    {
        FilePath = filePath;
    }

    public FileDataStore() : this(DefaultFilePath()) { }

    public bool HasPersistedData
    {
        get
        {
            try
            {
                var info = new FileInfo(FilePath);
                return info.Exists && info.Length > 0;
            }
            catch
            {
                return false;
            }
        }
    }

    /// <summary>
    /// Rounds every date in <paramref name="data"/> to whole seconds, matching
    /// what a save/load cycle produces. The engine stamps this on values it
    /// holds in memory so they stay equal to what is on disk.
    /// </summary>
    public static AppData NormalizingDates(AppData data) => data with
    {
        Reminders = data.Reminders.Select(reminder => reminder with
        {
            CreatedAt = reminder.CreatedAt.RoundedToSecond(),
            LastFiredAt = reminder.LastFiredAt?.RoundedToSecond(),
            LastAcknowledgedAt = reminder.LastAcknowledgedAt?.RoundedToSecond(),
            SnoozedUntil = reminder.SnoozedUntil?.RoundedToSecond(),
        }).ToList(),
        Events = data.Events.Select(e => e with
        {
            Date = e.Date.RoundedToSecond(),
        }).ToList(),
        Settings = data.Settings with
        {
            PausedUntil = data.Settings.PausedUntil?.RoundedToSecond(),
        },
    };

    /// <summary>
    /// Where an unreadable data file is preserved before the engine's fallback
    /// path overwrites it with defaults.
    /// </summary>
    public string CorruptBackupPath => Path.ChangeExtension(FilePath, "corrupt.json");

    public AppData Load()
    {
        lock (_gate)
        {
            if (!File.Exists(FilePath))
            {
                return new AppData { Reminders = DefaultReminders.StarterSet() };
            }
            var raw = File.ReadAllBytes(FilePath);
            if (raw.Length == 0)
            {
                return new AppData { Reminders = DefaultReminders.StarterSet() };
            }
            try
            {
                return AppDataJson.Decode(raw);
            }
            catch
            {
                // The engine reacts to an unreadable file by starting from the
                // defaults, and its next persist overwrites this file. Keep
                // the bytes around first, so a decode bug degrades into a
                // recoverable incident instead of silently destroying every
                // reminder the user configured.
                try
                {
                    File.Copy(FilePath, CorruptBackupPath, overwrite: true);
                }
                catch
                {
                    // Preserving the backup is best-effort, as on the Mac.
                }
                throw;
            }
        }
    }

    public void Save(AppData data)
    {
        lock (_gate)
        {
            var encoded = AppDataJson.Encode(NormalizingDates(data));
            var directory = Path.GetDirectoryName(FilePath);
            if (!string.IsNullOrEmpty(directory))
            {
                Directory.CreateDirectory(directory);
            }
            // Atomic write: a crash mid-save must not leave a truncated file
            // that loses every reminder the user configured.
            var temporary = FilePath + ".tmp";
            File.WriteAllBytes(temporary, encoded);
            File.Move(temporary, FilePath, overwrite: true);
        }
    }
}

/// <summary>An in-memory store for tests and previews.</summary>
public sealed class InMemoryDataStore : IDataStoring
{
    public AppData Data { get; private set; }
    public int SaveCount { get; private set; }

    public InMemoryDataStore(AppData? data = null)
    {
        Data = data ?? new AppData();
    }

    public bool HasPersistedData => SaveCount > 0;

    public AppData Load() => Data;

    public void Save(AppData data)
    {
        Data = data;
        SaveCount += 1;
    }
}

/// <summary>The reminders a brand-new install starts with.</summary>
public static class DefaultReminders
{
    /// <summary>
    /// Pressure relief and movement prompts, plus a couple of general examples
    /// so the concept of priority tiers is visible immediately.
    /// </summary>
    public static IReadOnlyList<Reminder> StarterSet(Instant? now = null)
    {
        var createdAt = now ?? SystemClock.Instance.GetCurrentInstant();
        return
        [
            new Reminder
            {
                Title = "Tilt Back",
                Message = "Tilt your chair back for 5 minutes. Stop working and listen to calming music.",
                Schedule = new Schedule.Interval(60),
                Priority = Priority.Critical,
                SymbolName = "figure.seated.side",
                ActivityDurationSeconds = 5 * 60,
                CreatedAt = createdAt,
            },
            new Reminder
            {
                Title = "Weight Shift",
                Message = "Activate your glutes and redistribute your weight.",
                Schedule = new Schedule.Interval(20),
                Priority = Priority.Subtle,
                SymbolName = "arrow.up.and.down.circle",
                CreatedAt = createdAt,
            },
            new Reminder
            {
                Title = "Drink Water",
                Message = "Have a drink of water.",
                Schedule = new Schedule.Interval(60),
                Priority = Priority.Normal,
                SymbolName = "drop.fill",
                CreatedAt = createdAt,
            },
            // Ships switched off so the user opts in rather than being
            // surprised by it.
            new Reminder
            {
                Title = "Stretch & Range of Motion",
                Message = "Run through your physio stretches.",
                Schedule = new Schedule.DailyAt(17, 0, 2),
                Priority = Priority.Important,
                SymbolName = "figure.flexibility",
                ActivityDurationSeconds = 10 * 60,
                IsEnabled = false,
                CreatedAt = createdAt,
            },
        ];
    }
}
